plugins/grpc-ecosystem/gateway/v2.19.1/separate_pkg_additional_imports.patch

```diff
@@ -11,102 +72,30 @@ index 1cfc4b02..11e1368d 100644
-+}
-+
-+// GetSeparatePackage returns separatePackage
-+func (r *Registry) GetSeparatePackage() bool {
-+	return r.separatePackage
-+}
-+
-+// SetAdditionalImports sets additionalImports
-+func (r *Registry) SetAdditionalImports(imports []string) {
-+	r.additionalImports = imports
-+}
-+
-+// GetAdditionalImports returns additionalImports
-+func (r *Registry) GetAdditionalImports() []string {
-+	return r.additionalImports
-+}
-+
- // SetPreserveRPCOrder sets preserveRPCOrder
- func (r *Registry) SetPreserveRPCOrder(preserve bool) {
- 	r.preserveRPCOrder = preserve
 diff --git a/internal/descriptor/services.go b/internal/descriptor/services.go
-index 0ff28d99..f5b409bc 100644
+index ad1764ce..ae58f7e3 100644
 --- a/internal/descriptor/services.go
 +++ b/internal/descriptor/services.go
-@@ -3,6 +3,7 @@ package descriptor
- import (
- 	"errors"
- 	"fmt"
-+	"path/filepath"
- 	"strings"
- 
- 	"github.com/grpc-ecosystem/grpc-gateway/v2/internal/httprule"
-@@ -20,6 +21,7 @@ func (r *Registry) loadServices(file *File) error {
- 		grpclog.Infof("Loading services from %s", file.GetName())
- 	}
- 	var svcs []*Service
-+	var additionalImports []string
- 	for _, sd := range file.GetService() {
- 		if grpclog.V(2) {
- 			grpclog.Infof("Registering %s", sd.GetName())
-@@ -29,6 +31,35 @@ func (r *Registry) loadServices(file *File) error {
+@@ -29,6 +29,7 @@ func (r *Registry) loadServices(file *File) error {
  			ServiceDescriptorProto: sd,
  			ForcePrefixedName:      r.standalone,
  		}
-+		if r.separatePackage {
-+			// when generating a separate package for the gateway, we need to generate an import statement
-+			// for the gRPC stubs that are no longer in the same package. This is done by adding the grpc
-+			// package to the additionalImports list. In order to prepare a valid import statement, we'll replace
-+			// the source package name, something like: ../pet/v1/v1petgateway with ../pet/v1/v1petgrpc
-+			const (
-+				baseTypePackageName    = "protocolbuffers"
-+				baseTypePackageSubPath = baseTypePackageName + "/go"
-+				grpcPackageName        = "grpc"
-+				grpcPackageSubPath     = grpcPackageName + "/go"
-+			)
-+			packageName := strings.TrimSuffix(svc.File.GoPkg.Name, "gateway") + grpcPackageName
-+			svc.GRPCFile = &File{
-+				GoPkg: GoPackage{
-+					// additionally, as the `go_package` option is passed through from the generator, and can only be
-+					// set the one time, without making major changes, we'll use the package name sent through the
-+					// options as a basis, and replace the source package name with the grpc package name.
-+					Path: strings.Replace(
-+						filepath.Join(svc.File.GoPkg.Path, packageName),
-+						baseTypePackageSubPath,
-+						grpcPackageSubPath,
-+						1,
-+					),
-+					Name: strings.Replace(packageName, baseTypePackageSubPath, grpcPackageSubPath, 1),
-+				},
-+			}
-+			additionalImports = append(additionalImports, svc.GRPCFile.GoPkg.Path)
-+		}
-+		r.SetAdditionalImports(additionalImports)
```

Comment:
The bug reported in #1200 was related to this bit of code. We are looping over all files and all services, but only adding additional imports based on the current service (and overriding any previously set ones). 

So if you have multiple proto packages with multiple services, you'd end up with an incorrect imports when they are added later on.